changes to stage_in.py to support transparent caching

foobob opened the pull request:
here is a fist pass on the transparent caching mods. This for s3 only. Needs updating for https or the other cache-able files. If we can get concurrence on approach, I'll work on cleaning this up.

rbcgh committed:
changes to stage_in.py to support transparent caching

jplzhan reviewed (changes requested):
Here are the changes that I would request be made:

Minor (formatting and syntax related):
- [ ] Fix inconsistency with `"` and `'` strings (ex. line 124). Strings should be using `'`.
- [ ] Remove tab in the middle of lines 207 and 208.
- [ ] Place new functions for caching within a separate class (for namespacing).
- [ ] Comment blocks should not be indented unless as a subsection to the rest of its contents. Said subsection should be indented with tabs.
- [ ] Fix inconsistency with camelcase vs underscores in variable names.
- [ ] Rename `localCachePath` to `cache_dir`.
- [ ] Rename `urnCacheableFilePath`, `fullCacheableFilePath`, and `fullCacheableSysFilePath` to shorter, less similar names, and ones which are more obvious about their use (ex. `fullCacheableSysFilePath` should probably have `symlink` in its name).

Major:
- [ ] Remove caching logic from `StageIn.stage_in_s3` and move it to before/after the actual download step within `main`. Add a new argument called `dest: str` to the function. In line 213, create the `dest` folder (if it does not already exist) and download the base file to `dest`. This change should be mirrored with other `StageIn.stage_in_http` functions.
- [ ] Avoid global, writable variables.
- [ ] Clean up argument parsing in `main`. If `-C` is specified, then `dest` (argument supplied to `StageIn.stage_in_s3`) should be the caching directory specified by `-C`. If it is **not** specified, keep it as the default value (the current return value of `Util.create_inputs_dir`, which is the default value of the optional `inputs_dir` argument).
- [ ] Rename `Util.create_inputs_dir` to `Util.create_dest`. Rename the `inputs_dir` argument to `dest`.
- [ ] Acquire the `cache_lock` before the general download commands within `main`, if `-C` is specified. Check for `cache_hit` and only call for `StageIn.stage_in_s3` if failed.
- [ ] Create the symlink if `dlpath` is downloaded, and if `-C is specified. Copy it to the original `inputs_dir` if `-s` is not specified.
- [ ] Divide into new functions if the code blocks added to main makes it too bulky.

rbcgh committed:
Reworked transparent caching from initial prototype to more general
implementation that can handle all download methods. specifics for MAAP
downloads that use 'collection_concept_id' still need to be worked in,
but the framework is there to query and cache in either the one or two
pass methods. Still working on the ETAG implementation but the basic
structure is there. Also cleaned up per comments.

foobob commented:
See updated version.

rbcgh committed:
This is the best I can do with merge conflicts without being able to test
anything.
Most things above main should be fine, and the basic structure in main
should be OK, but there are bound to be problems without any testing.

foobob commented:
I merged what I could without actually being able to test anything. most things above main should be OK, but main probably has some issues because its not been tested.

rbcgh committed:
This is the best I can do without having an example of the json input
file formats. It passes basic tests when I hard code parameters for a
single file.

rbcgh committed:
Merged and tested both s3 and http to handle json input and caching
dropped get_url, but cannot handle StageIn.stage_in_maap granules
without additional processing in main.

foobob commented:
I tried to merge but I don't have write permission to your repo. I pulled what you had at 6pm today and merged that all in with my changes.
